### PR TITLE
Fix `skip_special_tokens` for `Wav2Vec2CTCTokenizer._decode`

### LINKED ISSUE #29243: `skip_special_tokens` for `Wav2Vec2CTCTokenizer` does not work expectedly.
### System Info

- `transformers` version: 4.37.2
- Platform: Linux-5.15.0-1042-nvidia-x86_64-with-glibc2.35
- Python version: 3.10.13
- Huggingface_hub version: 0.20.1
- Safetensors version: 0.4.2
- Accelerate version: 0.26.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: DDP

### Who can help?

@sanchit-gandhi @ArthurZucker 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
processor = Wav2Vec2Processor.from_pretrained(model_args.model_name_or_path)
model = Wav2Vec2ConformerForCTC.from_pretrained(
    model_args.model_name_or_path, ctc_loss_reduction="mean", pad_token_id=processor.tokenizer.pad_token_id
)

metric_wer = evaluate.load("wer")
metric_cer = evaluate.load("cer")

def _compute_metrics(pred):
    pred.label_ids[pred.label_ids == -100] = processor.tokenizer.pad_token_id

    # ["<s>": 1, "</s>": 2, "<unk>": 3, "<pad>": 0]
    preds = processor.batch_decode(pred.predictions, skip_special_tokens=True)
    labels = processor.batch_decode(pred.label_ids, skip_special_tokens=True, group_tokens=False)

    preds = [postprocess(text) if len(text) != 0 else "-" for text in preds]
    labels = [postprocess(sentence) for sentence in labels]

    preds = [re.sub(r"\s+", "", text) for text in preds]
    labels = [re.sub(r"\s+", "", text) for text in labels]

    wer = 100 * metric_wer.compute(predictions=preds, references=labels)
    cer = 100 * metric_cer.compute(predictions=preds, references=labels)

    return {"wer": wer, "cer": cer}

def _preprocess_logits_for_metrics(logits, labels=None):
    return torch.argmax(logits, dim=-1)

trainer = Trainer(
    args=training_args,
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    data_collator=data_collator,
    tokenizer=processor,
    compute_metrics=_compute_metrics,
    preprocess_logits_for_metrics=_preprocess_logits_for_metrics,
)
```

### Expected behavior

I want to train ASR model and this issue came out when I evaluate during training.
`preds = processor.batch_decode(pred.predictions, skip_special_tokens=True)`'s results should remove all special tokens (`<s>`, `<\s>`, `<unk>`, `<pad>`). However, in my experiment, `<unk>` is not removed.

So, I look at the code for `Wav2Vec2CTCTokenizer._decode` (transformers / models / wav2vec2 / tokenization_wav2vec2.py).
```python
 def _decode(
        self,
        token_ids: List[int],
        skip_special_tokens: bool = False,
        clean_up_tokenization_spaces: bool = None,
        group_tokens: bool = True,
        spaces_between_special_tokens: bool = False,
        output_word_offsets: Optional[bool] = False,
        output_char_offsets: Optional[bool] = False,
    ) -> str:
        """
        special _decode function is needed for Wav2Vec2Tokenizer because added tokens should be treated exactly the
        same as tokens of the base vocabulary and therefore the function `convert_tokens_to_string` has to be called on
        the whole token list and not individually on added tokens
        """
        filtered_tokens = self.convert_ids_to_tokens(token_ids, skip_special_tokens=skip_special_tokens)

        result = []
        for token in filtered_tokens:
            if skip_special_tokens and token in self.all_special_ids:
                continue
            result.append(token)
```
This code removes the special tokens in the `filtered_tokens` by `self.all_special_ids`, but when I print the `filtered_tokens`, the outcome looks like `['|', '|', 'token1', 'token2', 'token3', '|', '|', 'token4', '|', '|', '|', 'token5', 'token6', '|', '|', '<unk>', '<unk>', '<unk>', '<unk>', '<unk>']`. 

Since `self.all_special_ids`'s elements are integers for special tokens, `if skip_special_tokens and token in self.all_special_ids:` statement does not work expectedly.

Shouldn’t it be `if skip_special_tokens and token in self.all_special_tokens:`?

### PR DESCRIPTION
**Pull request by msublee**
# What does this PR do?

Fixes #29243 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Do you mind adding a small test to make sure all special tokens are now skipped?

**Comment by msublee** [comment]
I'm unfamiliar with this kind of PR (to big repo). Where and How should I add a small test?

**Review by ArthurZucker — COMMENTED** [review]
There is a test here: https://github.com/huggingface/transformers/blob/d03de1c1070cee0fb7b9358669abc3f68ef56b75/tests/models/wav2vec2/test_tokenization_wav2vec2.py#L453

but it seems to not properly test so let's update it 😉

**Comment by msublee** [comment]
I updated the test!
Also, I excluded the `pad_token` filtering in `_decode` since `pad_token` is used as CTC-blank token and is filtered in `convert_tokens_to_string`.
https://github.com/msublee/transformers/blob/b27aa206ddf3fe66b36db587603141b3d0379a82/src/transformers/models/wav2vec2/tokenization_wav2vec2.py#L321
Finally, I applied this update to both `Wav2Vec2Tokenizer` and `Wav2Vec2CTCTokenizer`.

File: tests/models/wav2vec2/test_tokenization_wav2vec2.py [hunk 1]
```diff
@@ -145,8 +146,10 @@ def test_tokenizer_decode_added_tokens(self):
             [24, 22, 5, tokenizer.word_delimiter_token_id, 24, 22, 5, 77, tokenizer.pad_token_id, 34, 34],
         ]
         batch_tokens = tokenizer.batch_decode(sample_ids)
+        batch_tokens_2 = tokenizer.batch_decode(sample_ids, skip_special_tokens=True)
 
         self.assertEqual(batch_tokens, ["HELLO<unk>!?!?$$$", "BYE BYE<unk>$$$"])
+        self.assertEqual(batch_tokens_2, ["HELO!?!?", "BYE BYE"])
```
Comment (ArthurZucker) [inline tests/models/wav2vec2/test_tokenization_wav2vec2.py:152]: let's add a new token, like `"<new_tokens>"` and test that we can encode decode as we expect for both fast and slow tokenizers! 🤗

**Review by ArthurZucker — COMMENTED** [review]
Overall LGTM, we need to test a bit more, and I'll ask @sanchit-gandhi his expertise on this model and a second look 👀

File: tests/models/wav2vec2/test_tokenization_wav2vec2.py (hunk 1 above)
Comment (msublee) [reply tests/models/wav2vec2/test_tokenization_wav2vec2.py:152]: Is there a fast `Wav2Vec2CTCTokenizer`? I can't not find it!

Comment (ArthurZucker) [reply tests/models/wav2vec2/test_tokenization_wav2vec2.py:152]: Sorry there is no fast tokenizer here! my bad. Let's just use other tokens than `<unk>`!

Comment (msublee) [reply tests/models/wav2vec2/test_tokenization_wav2vec2.py:152]: Sorry for the late. I updated it.

**Review by ArthurZucker — APPROVED** [review]
Perfect thanks!

**Pull request merged**